### PR TITLE
[wip] Support serialized checkpoint loading

### LINKED ISSUE #9406: Support serialized checkpoint loading
### Motivation

Currently, all processes load the checkpoint at the same time. This can quickly OOM for large models. These use cases, especially with things like mixture of experts, might require serialized loading of checkpoint dicts across ranks (ie load the checkpoint one rank at a time per node). Could we enable this for DDP? 

Prior work: https://github.com/PyTorchLightning/pytorch-lightning/pull/8515

### Pitch

This would be controlled per training type plugin. Example pseudocode: https://gist.github.com/ananthsub/4ceedff56b2049a63bbb05ccd283b919

To work through: 
1. Should the TrainingTypePlugin have responsibility of calling `on_load_checkpoint` instead of the Trainer/connector? This would make sense as the TTP "owns" the LightningModule inside of the trainer, and since it already offers `load_model_state_dict`: https://github.com/PyTorchLightning/pytorch-lightning/blob/41ba639859cf6c6bf319eb33e5b3394504315962/pytorch_lightning/plugins/training_type/training_type_plugin.py#L159-L160

DeepSpeed already eschews most of the checkpoint connector logic when it comes to loading the lightning module state. This could be a gap for metrics, and this means we could be calling `on_load_checkpoint` multiple times with certain plugins. In my opinion, this points to all LightningModule state load/save/alterations should sit inside of the training type plugin. 

2. Does loading one process per node run into other issues with collectives potentially being used in `LightningModule.on_load_checkpoint` ?

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by jjenniferdai**
## What does this PR do?

Fixes #9406 

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 1]
```diff
@@ -153,11 +153,17 @@ def results(self) -> Optional[Union[_EVALUATE_OUTPUT, _PREDICT_OUTPUT]]:
...
+        self.on_load_checkpoint(checkpoint)
+        self.load_model_state_dict(checkpoint)
+        return checkpoint
 
     def load_model_state_dict(self, checkpoint: Mapping[str, Any]) -> None:
         self.lightning_module.load_state_dict(checkpoint["state_dict"])
 
+    def on_load_checkpoint(self, checkpoint: Mapping[str, Any]) -> None:
+        self.lightning_module.on_load_checkpoint(checkpoint)
+
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:166]: I don't think we should expose an `on_load_checkpoint` method separately here. instead, do you think the lightning module hook could be called directly in `TrainingTypePlugin.load_checkpoint` ?

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 2]
```diff
@@ -535,3 +536,13 @@ def teardown(self) -> None:
             self.lightning_module.cpu()
             # clean up memory
             torch.cuda.empty_cache()
+
+    def load_checkpoint(self, checkpoint_path: _PATH) -> Dict[str, Any]:
+        rank_zero_info(f"DistributedDataParallel has {self.num_processes} processes. Serializing to avoid CPU OOMs.")
+        for current_worker in range(self.num_processes):
+            if self.local_rank == current_worker and torch.distributed.is_initialized():
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/ddp.py:543]:
```suggestion
            if self.local_rank == current_worker:
```
we ought to fail loudly if torch distributed isn't initialized. otherwise nothing will be loaded

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -957,6 +954,8 @@ def _run(self, model: "pl.LightningModule") -> Optional[Union[_EVALUATE_OUTPUT,
 
         # check if we should delay restoring checkpoint till later
         if not self.accelerator.restore_checkpoint_after_pre_dispatch:
+            if self._ckpt_path:
+                self._load_checkpoint_weights()
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:958]: note to reviewer: this must be moved after the `setup_environment` is called to ensure that torch distributed is initialized. it also matches the symmetry in later on, and makes it easier for us to reason about merging these checkpoint loading paths altogether.

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 4]
```diff
@@ -154,12 +148,8 @@ def restore_model(self) -> None:
 
     def restore_model_weights(self, checkpoint_path: Optional[_PATH]) -> None:
         """Restore only the model weights."""
-        checkpoint = self._loaded_checkpoint
         if checkpoint_path is not None:
-            checkpoint = self._load_and_validate_checkpoint(checkpoint_path)
-
-        self.trainer.lightning_module.on_load_checkpoint(checkpoint)
-        self.trainer.training_type_plugin.load_model_state_dict(checkpoint)
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:162]: this means the state dict isn't loaded anymore. why this change?

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 5]
```diff
@@ -535,3 +536,13 @@ def teardown(self) -> None:
             self.lightning_module.cpu()
             # clean up memory
             torch.cuda.empty_cache()
+
+    def load_checkpoint(self, checkpoint_path: _PATH) -> Dict[str, Any]:
+        rank_zero_info(f"DistributedDataParallel has {self.num_processes} processes. Serializing to avoid CPU OOMs.")
+        for current_worker in range(self.num_processes):
+            if self.local_rank == current_worker:
+                checkpoint = super().load_checkpoint(checkpoint_path)
+                del checkpoint["state_dict"]
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/ddp.py:545]: load_model_state_dict should check if `"state_dict"` as a key is in the checkpoint dict. otherwise it'll fail on lookup

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 4 above)
Comment (jjenniferdai) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:162]: it gets called in new line 152 `_load_and_validate_checkpoint` --> `training_type_plugin.load_checkpoint` ?

File: pytorch_lightning/plugins/training_type/ddp.py (hunk 5 above)
Comment (jjenniferdai) [reply pytorch_lightning/plugins/training_type/ddp.py:545]: what's the expected behavior if `"state_dict"` key does not exist?

Comment (ananthsub) [reply pytorch_lightning/plugins/training_type/ddp.py:545]: got it!

Comment (ananthsub) [reply pytorch_lightning/plugins/training_type/ddp.py:545]: - the plugin can track whether it was the entity that deleted the key from the checkpoint in `load_checkpoint` . if it did, and there's no `"state_dict"` found when calling `load_model_state_dict` the plugin can simply return early
- if it didn't, then the plugin could raise an exception about the key not found as this is unexpected

Comment (jjenniferdai) [reply pytorch_lightning/plugins/training_type/ddp.py:545]: hm do you mean something like the below edit?

Comment (ananthsub) [reply pytorch_lightning/plugins/training_type/ddp.py:545]: yup, but this doesn't need to be on the base training type plugin. it can be initialized on the plugins which actually serialize checkpoints

**Comment by jjenniferdai** [comment]
messed up git, moving to #9605
